src/MongoDB.Driver.Core/Core/Authentication/SaslPrepHelper.cs

```diff
@@ -0,0 +1,806 @@
+        internal const int SurrogateMinCodepoint = 0x00d800;
+        internal const int SurrogateMaxCodePoint = 0x00dfff;
+        
+        /// <summary>
+        /// Return the SASLPrep-canonicalised version of the given <paramref name="str"/> for use as a query string.                          
+        /// This implements the {@code SASLPrep} algorithm defined in <a href="https://tools.ietf.org/html/rfc4013">RFC 4013</a>.          
+        /// See <a href="https://tools.ietf.org/html/rfc3454#section-7">RFC 3454, Section 7</a> for discussion of what a
+        /// query string is.
+        /// String normalization step in the .NET Standard version of the driver is skipped due to a lack of a string
+        /// normalization function.
+        /// </summary>
+        /// <param name="str">The string to canonicalise.</param>
+        /// <returns>The canonicalised string.</returns>
+        internal static string SaslPrepQuery(string str)
+        {
+            return SaslPrep(str, true);
+        }
+
+        /// <summary>
+        /// Return the SASLPrep-canonicalised version of the given <paramref name="str"/> for use as a stored string.                          
+        /// This implements the SASLPrep algorithm defined in <a href="https://tools.ietf.org/html/rfc4013">RFC 4013</a>.
+        /// See <a href="https://tools.ietf.org/html/rfc3454#section-7">RFC 3454, Section 7</a> for discussion of what a
+        /// stored string is.
+        /// String normalization step in the .NET Standard version of the driver is skipped due to a lack of a string
+        /// normalization function.
+        /// </summary>
+        ///<param name="str">The string to canonicalise.</param>
+        ///<returns>The canonicalised string.</returns>
+        internal static string SaslPrepStored(string str)
+        {
+            return SaslPrep(str, false);
+        }
+
+        private static string SaslPrep(string str, bool allowUnassigned) {
+            var chars = str.ToCharArray();
+
+            // 1. Map
+            // non-ASCII space chars mapped to space
+            for (var i = 0; i < str.Length; i++)
+            {
+                var ch = str[i];
+                if (NonAsciiSpace(ch)) 
+                {
+                    chars[i] = ' ';
+                }
+            }
+
+            var length = 0;
+            for (var i = 0; i < str.Length; i++)
+            {
+                var ch = chars[i];
+                if (!MappedToNothing(ch))
+                {
+                    chars[length++] = ch;
+                }
+            }
+
+            // 2. Normalize
+#if NET45
+            var normalized = new string(chars.Take(length).ToArray()).Normalize(NormalizationForm.FormKC);
+#else
+            // String normalization step in the .NET Standard version of the driver is skipped due to a lack of a string
+            // normalization function.
+            var normalized = new string(chars.Take(length).ToArray());
+#endif
```

Comment:
It would probably be a good idea to move what's common to both cases outside the #if.

Reply:
Fixed.